### ISSUE TITLE
GroupBy shouldn't use equalTo

### ISSUE THREAD

**Issue by haozhun** [prestodb/presto#3568]
This query fails because `Type.equalTo` is used instead of `is distinct from` in group by. However, `is distinct from` as is implemented today is not an operator.

```
select x
from (values (array[1], 0), (array[2], 0), (array[cast (null as bigint)], 0)) t(x,y)
group by x;
```

**Comment by raghavsethi** [comment]
Actually, we already have ```IS DISTINCT FROM``` semantics for ```GROUP BY```. The issue here is that we fail on computing hashes for arrays which contain null values.

**Comment by raghavsethi** [comment]
This is turning out to be more complicated than it appears. The main issues are:
- We need to be able to deal with boxed arguments to scalar functions
- The function needs to apply recursively, and therefore needs to be implemented for all complex types

**Comment by raghavsethi** [comment]
Fixing this properly would also fix #3569

**Comment by haozhun** [comment]
Fixing #3569 will be a step toward fixing this. This task kind of depend on #3569

**Comment by raghavsethi** [comment]
Also related to #4075.

**Comment by haozhun** [comment]
Fixed in #10392 by @elonazoulay
